[JSC] Add fast path for Object.assign(target, NamedProperties + IndexedProperties)

hyjorc1 opened the pull request:
#### 3e8cc7706222b0efa402b2bf003729c3913101e7
<pre>
[JSC] Add fast path for Object.assign(target, NamedProperties + IndexedProperties)
<a href="https://bugs.webkit.org/show_bug.cgi?id=261219">https://bugs.webkit.org/show_bug.cgi?id=261219</a>
rdar://115070621

Reviewed by Yusuke Suzuki.

Previously, we have fast path only for Object.assign(target, NamedProperties).
This patch extends the Object.assign fast path for source consists of
NamedProperties and IndexedProperties.

                                                   without                     with

object-assign-slow-put-array-storage-source     9.3211+-0.0820     ^      4.7620+-0.0325        ^ definitely 1.9574x faster
object-assign-contiguous-source                 6.5756+-0.2071     ^      2.4984+-0.0273        ^ definitely 2.6320x faster
object-assign-int32-source                      6.4450+-0.0493     ^      2.6511+-0.0796        ^ definitely 2.4310x faster
object-assign-multiple-sources                 22.1794+-0.1457     ^     10.5955+-0.2918        ^ definitely 2.0933x faster
object-assign-array-storage-source              9.2886+-0.0758     ^      4.7451+-0.2034        ^ definitely 1.9575x faster
object-assign-double-source                     6.5618+-0.1262     ^      2.8638+-0.0560        ^ definitely 2.2913x faster
object-assign-empty-target-sp3-charts-chartjs   2.8745+-0.0313     ^      1.8253+-0.0134        ^ definitely 1.5748x faster

* JSTests/microbenchmarks/object-assign-empty-target-array-storage-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-contiguous-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-double-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-int32-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-slow-put-array-storage-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-sp3-charts.js: Added.
* JSTests/stress/object-assign-empty-indexed-property-source.js: Added.
(shouldBe):
(test):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::copyArrayStorage):
(JSC::JSObject::putOwnDataPropertyBatching):
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::isEmptyJSFinalObject):
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::fastForEachPropertyWithSideEffectFreeFunctor):
* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectAssignFast):
* Source/JavaScriptCore/runtime/Structure.h:
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::canPerformFastPropertyEnumeration const):

Canonical link: <a href="https://commits.webkit.org/267797@main">https://commits.webkit.org/267797@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c7b5cbb8f59e71ebfb099251f0cabe2d870aaae

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18013 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18546 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19506 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16539 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21302 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18161 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18617 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17896 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18190 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15359 "2 api tests failed or timed out") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20361 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15424 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16105 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22690 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15287 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16434 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16274 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20549 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/16894 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16846 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14264 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/35/builds/19259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15962 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/35/builds/19259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4224 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20325 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/20492 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16690 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4684 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/StructureInlines.h:
```diff
@@ -832,12 +832,10 @@ ALWAYS_INLINE bool Structure::canPerformFastPropertyEnumeration() const
     // https://bugs.webkit.org/show_bug.cgi?id=185358
 
-    if (hasIndexedProperties(indexingType()))
+    if (!isTargetEmptyJSFinalObject && hasIndexedProperties(indexingType()))
```
Why? We should always allow this.

hyjorc1 replied:
Updated. `canPerformFastPropertyEnumeration` is used in JSONObject.cpp. Maybe we can update it in the future patches. Extract `canPerformFastPropertyEnumerationCommon`.

Constellation commented on Source/JavaScriptCore/runtime/JSObject.cpp:
```diff
@@ -3196,6 +3196,79 @@ bool JSObject::putByIndexBeyondVectorLengthWithArrayStorage(JSGlobalObject* glob
     return true;
 }
 
+void JSObject::copyArrayStorage(JSGlobalObject* globalObject, JSObject* other)
```
Contiguous etc., is not `ArrayStorage`, so this name looks wrong.

hyjorc1 replied:
Removed. Use `JSObject::forEachIndexedProperty` instead.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -387,7 +387,9 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
             }
 
-            if (objectAssignFast(vm, target, source, properties, values))
+            bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+            RETURN_IF_EXCEPTION(scope, { });
+            if (objectAssignFastSucceeded)
```
Why do we need this change?

hyjorc1 replied:
In `JSObject::copyArrayStorage`, `SparseArrayValueMap:: putEntry` might throw exception. 
```
                map->putEntry(globalObject, this, it->key, value, false);
                RETURN_IF_EXCEPTION(scope, void());
```

hyjorc1 replied:
Updated. Now, use `putDirectIndex` instead and it may calls `putByIndexBeyondVectorLengthWithoutAttributes` which might throw exceptions.

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -311,7 +311,9 @@ JSC_DEFINE_JIT_OPERATION(operationObjectAssignObject, void, (JSGlobalObject* glo
         // Do not clear since Vector::clear shrinks the backing store.
-        if (objectAssignFast(vm, target, source, properties, values))
+        bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+        RETURN_IF_EXCEPTION(scope, void());
```
You should use EXCEPTION_ASSERT instead.

Constellation replied:
Ignore this, probably this is fine.

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -342,7 +344,9 @@ JSC_DEFINE_JIT_OPERATION(operationObjectAssignUntyped, void, (JSGlobalObject* gl
         Vector<RefPtr<UniquedStringImpl>, 8> properties;
         MarkedArgumentBuffer values;
-        if (objectAssignFast(vm, target, source, properties, values))
+        bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+        RETURN_IF_EXCEPTION(scope, void());
+        if (objectAssignFastSucceeded)
```
You should use EXCEPTION_ASSERT instead.

Constellation replied:
Ignore this, probably this is fine.

Constellation commented on Source/JavaScriptCore/runtime/JSObjectInlines.h:
```diff
@@ -868,13 +868,71 @@ bool JSObject::fastForEachPropertyWithSideEffectFreeFunctor(VM& vm, const Functo
+            functor(i, JSValue(JSValue::EncodeAsDouble, value));
+        }
+        break;
+    }
```
This causes GC issues as I said.

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -311,7 +311,9 @@ JSC_DEFINE_JIT_OPERATION(operationObjectAssignObject, void, (JSGlobalObject* glo
         // Do not clear since Vector::clear shrinks the backing store.
-        if (objectAssignFast(vm, target, source, properties, values))
+        bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+        EXCEPTION_ASSERT(!scope.exception()); // We should not throw any exception here.
```
For now, let's just use RETURN_IF_EXCEPTION. It is possible that objectAssignFast can throw an exception in particular due to indexed properties set.

Constellation commented on Source/JavaScriptCore/dfg/DFGOperations.cpp:
```diff
@@ -342,7 +344,9 @@ JSC_DEFINE_JIT_OPERATION(operationObjectAssignUntyped, void, (JSGlobalObject* gl
-        if (objectAssignFast(vm, target, source, properties, values))
+        bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+        EXCEPTION_ASSERT(!scope.exception()); // We should not throw any exception here.
+        if (objectAssignFastSucceeded)
```
Ditto.

Constellation commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -203,6 +206,31 @@ class JSObject : public JSCell {
+            unsigned count = std::min(storage->length(), storage->vectorLength());
+            if (SparseArrayValueMap* map = storage->m_sparseMap.get())
+                count += map->size();
+            return count;
```
This is not correct since SparseArrayValueMap is not contiguously holding properties.

Constellation replied:
So, just return bool. And don't do these counting.
Do exactly,
```
ArrayStorage* storage = m_butterfly->arrayStorage();
unsigned count = std::min(storage->length(), storage->vectorLength());
if (count)
    return true;
SparseArrayValueMap* map = storage->m_sparseMap.get();
if (!map)
    return false;
return map->size();
```

Constellation commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -203,6 +206,31 @@ class JSObject : public JSCell {
         return m_butterfly->vectorLength();
     }
     
+    unsigned getIndexedPropertyCount() const
```
`getIndexedPropertyCount` name is not correct. When returning publicLength from Int32Shape for example, it does not mean that it has publicLength count of properties. Some of them can be an array hole.
This function should just return bool whether it has existing indexed properties.
Like `bool canHaveExistingOwnIndexedProperties() const`

Constellation commented on Source/JavaScriptCore/runtime/JSObjectInlines.h:
```diff
@@ -868,13 +868,70 @@ bool JSObject::fastForEachPropertyWithSideEffectFreeFunctor(VM& vm, const Functo
+        if (SparseArrayValueMap* map = storage->m_sparseMap.get()) {
+            if (map->begin() == map->end())
+                return;
+            for (SparseArrayValueMap::const_iterator it = map->begin(); it != map->end(); ++it) {
```
Let's just use,
```
for (auto& [key, value] : *map) {
    properties.append(key);
    values.appendWithCrashOnOverflow(value.getNonSparseMode());
}
```

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
```diff
@@ -66,6 +67,16 @@ ALWAYS_INLINE bool objectAssignFast(VM& vm, JSObject* target, JSObject* source,
+    bool targetHasIndexedOrIsEmpty = hasIndexedOrIsEmpty(target);
+    unsigned sourceIndexedPropertyCount = source->getIndexedPropertyCount();
+    if (sourceIndexedPropertyCount && !targetHasIndexedOrIsEmpty)
+        return false;
```
Again, hasIndexedOrIsEmpty is not necessary as I said. Remove these lines.

Constellation commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -1619,6 +1647,18 @@ inline bool JSObject::putDirect(VM& vm, PropertyName propertyName, JSValue value
+{
+    return hasIndexedProperties(object->indexingType()) || isEmptyJSFinalObject(object);
+}
+
```
Remove this.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -339,7 +339,13 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
+
+            // If the source has indexed properties, then ideally we want to do the fast path when
+            // the target also has indexed properties or is an empty JSFinalObject.
+            bool targetHasIndexedOrIsEmpty = hasIndexedOrIsEmpty(target);
```
Again, as I said, this is not necessary. Remove this.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -339,7 +339,13 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
+            bool targetHasIndexedOrIsEmpty = hasIndexedOrIsEmpty(target);
+            unsigned sourceIndexedPropertyCount = source->getIndexedPropertyCount();
+            bool cannotAssignSourceToTargetFast = sourceIndexedPropertyCount && !targetHasIndexedOrIsEmpty;
+            if (!source->staticPropertiesReified() || !source->structure()->canPerformFastPropertyEnumerationCommon() || cannotAssignSourceToTargetFast) {
```
For this `willBatch` thing, it is totally fine that you can just skip indexed properties case as before. Just do that.
The rest is still using fast path. Check the previous condition here. Make `willBatch = false` when source has indexed properties here.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -350,6 +356,11 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
+                source->forEachIndexedProperty(globalObject, [&](unsigned index, JSValue value) {
+                    target->putDirectIndex(globalObject, index, value);
+                    RETURN_IF_EXCEPTION(scope, void());
+                });
```
Remove this.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -387,7 +398,9 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
-            if (objectAssignFast(vm, target, source, properties, values))
+            bool objectAssignFastSucceeded = objectAssignFast(globalObject, target, source, properties, values);
+            EXCEPTION_ASSERT(!scope.exception()); // We should not throw any exception here.
+            if (objectAssignFastSucceeded)
```
Let this handle indexed one instead.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -465,30 +485,47 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorEntries, (JSGlobalObject* globalObject
+                {
+                    ObjectInitializationScope initializationScope(vm);
+                    if (LIKELY(entry = JSArray::tryCreateUninitializedRestricted(initializationScope, nullptr, globalObject->arrayStructureForIndexingTypeDuringAllocation(ArrayWithContiguous), 2))) {
+                        entry->initializeIndex(initializationScope, 0, JSValue(indexedProperties[i]));
```
This is wrong. entries' key should be string.
But this patch has so many issues on entries implementation. Don't change this, and just focus on Object.assign.
Let's just use slow path for indexed properties as before.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -443,19 +463,19 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorEntries, (JSGlobalObject* globalObject
             }
 
-            if (!cachedButterfly && properties.size() < MIN_SPARSE_ARRAY_INDEX && !globalObject->isHavingABadTime() && targetStructure->canCacheOwnPropertyNames()) {
+            if (!cachedButterfly && namedProperties.size() < MIN_SPARSE_ARRAY_INDEX && !globalObject->isHavingABadTime() && targetStructure->canCacheOwnPropertyNames()) {
```
This is wrong. We should not cache when there is indexed properties.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -415,24 +428,31 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorEntries, (JSGlobalObject* globalObject
-        MarkedArgumentBuffer values;
+        Vector<RefPtr<UniquedStringImpl>, 8> namedProperties;
+        MarkedArgumentBuffer namedPropertyValues;
         bool canUseFastPath = target->fastForEachPropertyWithSideEffectFreeFunctor(vm, [&](const PropertyTableEntry& entry) -> bool {
```
Let's not change this `objectConstructorEntries`. Just check indexed properties as before, and go to the slow path. The current change in this patch has many issues. Let's focus on Object.assign.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
```diff
@@ -84,6 +95,11 @@ ALWAYS_INLINE bool objectAssignFast(VM& vm, JSObject* target, JSObject* source,
 
+    source->forEachIndexedProperty(globalObject, [&](unsigned index, JSValue value) {
+        target->putDirectIndex(globalObject, index, value);
+        EXCEPTION_ASSERT_UNUSED(scope, !scope.exception());
```
This is not correct. Exception can happen (OOM).
You need to make `forEachIndexedProperty` stoppable in the middle of iteration. See IterationStatus usage in the other functions.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
```diff
@@ -84,6 +95,11 @@ ALWAYS_INLINE bool objectAssignFast(VM& vm, JSObject* target, JSObject* source,
+    source->forEachIndexedProperty(globalObject, [&](unsigned index, JSValue value) {
+        target->putDirectIndex(globalObject, index, value);
+        EXCEPTION_ASSERT_UNUSED(scope, !scope.exception());
+    });
```
You should check exception here too.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
```diff
@@ -84,6 +95,11 @@ ALWAYS_INLINE bool objectAssignFast(VM& vm, JSObject* target, JSObject* source,
     if (!canUseFastPath)
         return false;
 
+    source->forEachIndexedProperty(globalObject, [&](unsigned index, JSValue value) {
```
I don't think spreading code here is the right idea.
As I said, please put this out of this function. And do quick check here.
```
if (object->canHaveExistingIndexedProperties()) {
    objectAssignIndexedPropertiesFast(target, source);
    RETURN_IF_EXCEPTION(scope, { });
}
```

Constellation commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -203,6 +206,31 @@ class JSObject : public JSCell {
+    unsigned getIndexedPropertyCount() const
+    {
+        if (!hasIndexedProperties(indexingType()))
+            return 0;
```
return false;

Constellation commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -203,6 +206,31 @@ class JSObject : public JSCell {
+        switch (indexingType()) {
+        case ALL_BLANK_INDEXING_TYPES:
+        case ALL_UNDECIDED_INDEXING_TYPES:
+            return 0;
```
return false;

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -417,18 +419,21 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorEntries, (JSGlobalObject* globalObject
-        bool canUseFastPath = target->fastForEachPropertyWithSideEffectFreeFunctor(vm, [&](const PropertyTableEntry& entry) -> bool {
-            if (entry.attributes() & PropertyAttribute::DontEnum)
-                return true;
+        bool canUseFastPath = false;
```
It becomes always false, right?

hyjorc1 replied:
Oops, it should be
```
        bool canUseFastPath = false;
        if (!target->canHaveExistingOwnIndexedProperties()) {
            canUseFastPath = target->fastForEachPropertyWithSideEffectFreeFunctor(vm, [&](const 
```

Constellation commented on Source/JavaScriptCore/runtime/JSObject.h:
```diff
@@ -203,6 +206,34 @@ class JSObject : public JSCell {
     }
     
+    unsigned canHaveExistingOwnIndexedProperties() const
+    {
```
Let's add a comment saying that this is only valid after using `canPerformFastPropertyEnumerationCommon`. This code is not checking getOwnPropertySlot override etc.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
```diff
@@ -50,7 +50,64 @@ ALWAYS_INLINE bool canPerformFastPropertyNameEnumerationForJSONStringifyWithSide
-ALWAYS_INLINE bool objectAssignFast(VM& vm, JSObject* target, JSObject* source, Vector<RefPtr<UniquedStringImpl>, 8>& properties, MarkedArgumentBuffer& values)
+ALWAYS_INLINE void objectAssignIndexedPropertiesFast(JSGlobalObject* globalObject, JSObject* target, JSObject* source)
+{
+    VM& vm = globalObject->vm();
```
Use `forEachIndexedProperty` to implement `objectAssignIndexedPropertiesFast`.
Stop iteration when exceptions happens with `IterationStatus::Done`.

Constellation commented on Source/JavaScriptCore/runtime/JSObjectInlines.h:
```diff
@@ -868,13 +868,69 @@ bool JSObject::fastForEachPropertyWithSideEffectFreeFunctor(VM& vm, const Functo
+    case ALL_DOUBLE_INDEXING_TYPES: {
+        unsigned usedLength = m_butterfly->publicLength();
+        for (unsigned i = 0; i < usedLength; ++i) {
+            JSValue value = getDirectIndex(globalObject, i);
```
This can cause an exception. So you need to check exception after this.

Constellation commented on Source/JavaScriptCore/runtime/JSObjectInlines.h:
```diff
@@ -868,13 +868,69 @@ bool JSObject::fastForEachPropertyWithSideEffectFreeFunctor(VM& vm, const Functo
 
+template<typename Functor>
+void JSObject::forEachIndexedProperty(JSGlobalObject* globalObject, const Functor& functor)
+{
```
Add `ASSERT(structure()->canPerformFastPropertyEnumerationCommon());` This should be called only when it is true.

Constellation commented on Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
```diff
@@ -561,26 +566,34 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorValues, (JSGlobalObject* globalObject,
+            target->forEachIndexedProperty(globalObject, [&](unsigned, JSValue value) {
+                indexedPropertyValues.appendWithCrashOnOverflow(value);
+                return IterationStatus::Continue;
+            });
```
Exception check is necessary.

Constellation commented on Source/JavaScriptCore/runtime/JSObjectInlines.h:
```diff
@@ -868,13 +868,74 @@ bool JSObject::fastForEachPropertyWithSideEffectFreeFunctor(VM& vm, const Functo
+{
+    ASSERT(structure()->canPerformFastPropertyEnumerationCommon());
+    if (!canHaveExistingOwnIndexedProperties())
+        return;
```
Let's remove this. If we would like to early return here, this should be placed in the caller side of this function.

Constellation reviewed (approved):
r=me with some comments.

hyjorc1 committed:
[JSC] Add fast path for Object.assign(target, NamedProperties + IndexedProperties)
https://bugs.webkit.org/show_bug.cgi?id=261219
rdar://115070621

Reviewed by Yusuke Suzuki.

Previously, we have fast path only for Object.assign(target, NamedProperties).
This patch extends the Object.assign fast path for source consists of
NamedProperties and IndexedProperties.

                                                   without                     with

object-assign-slow-put-array-storage-source     9.3211+-0.0820     ^      4.7620+-0.0325        ^ definitely 1.9574x faster
object-assign-contiguous-source                 6.5756+-0.2071     ^      2.4984+-0.0273        ^ definitely 2.6320x faster
object-assign-int32-source                      6.4450+-0.0493     ^      2.6511+-0.0796        ^ definitely 2.4310x faster
object-assign-multiple-sources                 22.1794+-0.1457     ^     10.5955+-0.2918        ^ definitely 2.0933x faster
object-assign-array-storage-source              9.2886+-0.0758     ^      4.7451+-0.2034        ^ definitely 1.9575x faster
object-assign-double-source                     6.5618+-0.1262     ^      2.8638+-0.0560        ^ definitely 2.2913x faster
object-assign-empty-target-sp3-charts-chartjs   2.8745+-0.0313     ^      1.8253+-0.0134        ^ definitely 1.5748x faster

* JSTests/microbenchmarks/object-assign-empty-target-array-storage-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-contiguous-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-double-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-int32-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-slow-put-array-storage-source.js: Added.
* JSTests/microbenchmarks/object-assign-empty-target-sp3-charts.js: Added.
* JSTests/stress/object-assign-empty-indexed-property-source.js: Added.
(shouldBe):
(test):
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
(JSC::DFG::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/runtime/JSObject.cpp:
(JSC::JSObject::copyArrayStorage):
(JSC::JSObject::putOwnDataPropertyBatching):
* Source/JavaScriptCore/runtime/JSObject.h:
(JSC::isEmptyJSFinalObject):
* Source/JavaScriptCore/runtime/JSObjectInlines.h:
(JSC::JSObject::fastForEachPropertyWithSideEffectFreeFunctor):
* Source/JavaScriptCore/runtime/ObjectConstructor.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/ObjectConstructorInlines.h:
(JSC::objectAssignFast):
* Source/JavaScriptCore/runtime/Structure.h:
* Source/JavaScriptCore/runtime/StructureInlines.h:
(JSC::Structure::canPerformFastPropertyEnumeration const):

Canonical link: https://commits.webkit.org/267797@main
